[Issue 452] Migrate Variable Registration to Enable XML Recorder Recording of All Events for a Single Neuron

xiangli6 opened the pull request:
Fixes Issue #452 

1.Remove all concrete classes in the XML Recorder except for the XmlRecorder class. 
2. Change the recorder name to XmlRecorder in the related config files
3. Refactor the XmlRecorder class. Currently, this class only records Eventbuffer variable and it stores all events of a single neuron (neuron 7).

stiber commented on Simulator/Recorders/Hdf5Recorder.h:
```diff
@@ -70,6 +70,8 @@ class Hdf5Recorder : public IRecorder {
    /// Registered to OperationManager as Operation::printParameters
    virtual void printParameters() override;
 
+   virtual void registerVariables(std::string varName, EventBuffer &recordVar);
```
Since this is just registering one variable at a time, I think it could be called `registerVariable()`

stiber replied:
Also, should this be declared `override`?

xiangli6 replied:
Yes, I should do that.

stiber commented on Simulator/Recorders/IRecorder.h:
I think this class should be changed to `Recorder`.

xiangli6 replied:
Will do it

stiber commented on Simulator/Recorders/Neuro/Hdf5GrowthRecorder.h:
This class will go away, right?

xiangli6 replied:
This issue is only applied to XML Recorder. I will open another issue next time for HDF5 Recorder changes, then delete the hdf5 related classes in the that issue. Will this work?

stiber replied:
Yes, that sounds good. Just wanted to make sure it wasn't something overlooked here.

stiber commented on Simulator/Recorders/XmlRecorder.h:
```diff
@@ -57,16 +56,34 @@ class XmlRecorder : public IRecorder {
+   void registerVariables(string varName, EventBuffer &recordVar) override;
+
 protected:
+   // a signle neuron number
+   string single_neuron_name;
+
+   // all events for a single neuron
+   EventBuffer *variable_first;
```
These don't meet our requirements for member variable name styles. Also, needs more comments. I know these are just stepping stones along the way, but good to get in the habit, and most of the commenting will be reusable.

stiber commented on Simulator/Recorders/XmlRecorder.cpp:
```diff
@@ -80,94 +76,45 @@ void XmlRecorder::term()
-      getStarterNeuronMatrix(starterNeurons, layout->starterMap_);
-      resultOut_ << "   " << starterNeurons.toXML("starterNeurons") << endl;
+/// convert internal buffer to XML string
+string XmlRecorder::toXML(string name, vector<uint64_t> single_neuron_buffer) const
```
We talked about the XML file formatting needed.

stiber commented on Simulator/Vertices/Neuro/AllSpikingNeurons.cpp:
```diff
@@ -17,8 +18,12 @@ void AllSpikingNeurons::setupVertices()
                                     * Simulator::getInstance().getMaxFiringRate());
 
-   hasFired_.assign(size_, false);
-   vertexEvents_.assign(size_, maxSpikes);
+   hasFired_.resize(size_, false);
+   vertexEvents_.resize(size_, maxSpikes);
```
It's possible that @KamathScience  decided on `assign()`, rather than `resize()`; I'm not sure — consult with her.

stiber commented on Simulator/Vertices/Neuro/AllSpikingNeurons.h:
```diff
@@ -33,7 +33,7 @@ struct AllSpikingNeuronsDeviceProperties;
 class AllSpikingNeurons : public AllVertices {
 public:
-   AllSpikingNeurons() = default;
+   //AllSpikingNeurons() = default;
```
Why?

stiber reviewed (changes requested):
Comments on what has been done so far. We've already talked about some/much of this.

stiber commented on Simulator/Vertices/Neuro/AllSpikingNeurons.cpp:
```diff
@@ -19,6 +20,10 @@ void AllSpikingNeurons::setupVertices()
    vertexEvents_.assign(size_, maxSpikes);
+
+   // register variables
+   Recorder &recorder = *Simulator::getInstance().getModel()->getRecorder();
```
OK, hate to do this, but I know that @KamathScience has transitioned a lot of the `get*()` methods to return references, rather than pointers. I don't know if that applies here, or if it has made its way into documentation (though that would almost certainly not be the `gh-pages` documentation, yes). Consult with her about this.

stiber reviewed (changes requested):
Not sure if you're ready to merge quite yet. Re-request my review as needed. When you've done enough manual testing (or updated the regression tests) and are ready to merge, let me know explicitly and I will override the block due to not all tests passing, if needed.

xiangli6 commented:
> Not sure if you're ready to merge quite yet. Re-request my review as needed. When you've done enough manual testing (or updated the regression tests) and are ready to merge, let me know explicitly and I will override the block due to not all tests passing, if needed.

It is not ready to merge. I discussed this with Divya. My plan is starting to write a unit testing file for the new XmlRecorder class from next week. After that I will update the regression test files to pass the regression test. Will this work?

stiber commented on CMakeLists.txt:
I know this file will change, based on our conversation today.

xiangli6 replied:
I made some new changes 
1.  Converted the pointer in the XmlRecorder to a shared pointer and assigned the address of recordVar to singleNeuronEvents_:
         In the registerVariable function, instead of using std::make_shared, I assign the 
         address of recordVar to singleNeuronEvents_ by using:
                            singleNeuronEvents_ = std::shared_ptr<EventBuffer>(&recordVar, [](EventBuffer *) {});
         The reason is that if I create the shared_ptr<EventBuffer> using std::make_shared<EventBuffer>(recordVar), it 
         creates a new EventBuffer object and copies the data from recordVar into the newly created object. The 
         bufferPtr then holds the address of this newly created object instead of the address of the object that passed 
         into this method.
2. Removed ifdef UNIT_TEST in the XmlRecorder.h and deleted relative definition/code in the CMakeList file
3. Changed the XmlRecorderTest and CmakeList file so that when we test the XmlRecorder, we use  ./tests instead of creating a new recorderTest target (like serialTest test) in the CMakeList.txt.

stiber reviewed (changes requested):
I know this is still in progress; this will hopefully just catch me up so that I only see changes after this.

xiangli6 commented:
> I know this is still in progress; this will hopefully just catch me up so that I only see changes after this.
New changes:
1. Converted the pointer in the XmlRecorder to a shared pointer and assigned the address of recordVar to singleNeuronEvents_:
        In the registerVariable function, instead of using std::make_shared, I assign the address of recordVar to 
singleNeuronEvents_ by using:
                     singleNeuronEvents_ = std::shared_ptr<EventBuffer>(&recordVar, [](EventBuffer *) {});
        The reason is that if I create the shared_ptr<EventBuffer> using std::make_shared<EventBuffer>(recordVar), it creates a 
new EventBuffer object and copies the data from recordVar into the newly created object. The bufferPtr then holds the address of this newly created object instead of the address of the object that passed into this method.

2. Removed ifdef UNIT_TEST in the XmlRecorder.h and deleted relative definition/code in the CMakeList file

3. Changed the XmlRecorderTest and CmakeList file so that when we test the XmlRecorder, we use  ./tests instead of creating a new recorderTest target (like serialTest test) in the CMakeList.txt.

stiber commented on Simulator/Recorders/XmlRecorder.h:
```diff
@@ -57,16 +55,61 @@ class XmlRecorder : public IRecorder {
+   shared_ptr<EventBuffer> singleNeuronEvents_;
+
+   // history of accumulated event for a single neuron
+   std::vector<uint64_t> single_neuron_History_;
```
Why use the `_` naming style for this one member variable?

xiangli6 replied:
I should name it with singleNeuronHistory_. Sorry for this mistake

stiber reviewed (changes requested):
I'm back! And only one minor thing...

stiber reviewed (approved):
Good to go, once GPU testing shows not catastrophic problems.
